streaming/base/world.py

```diff
@@ -37,28 +40,109 @@ class World:
+        self.worker_of_node = self.worker % (ranks_per_node * workers_per_rank)
+        self.workers_per_node = ranks_per_node * workers_per_rank
+        self.worker_of_rank = self.worker % workers_per_rank
+        self.workers_per_rank = workers_per_rank
+        self.is_leader = not worker
+        self.is_local_leader = not self.worker_of_node
+
+    def to_json(self) -> Dict[str, Any]:
+        """Get a JSON version of this config.
+
+        Returns:
+            Dict[str, Any]: JSON config.
+        """
+        return dict(self.__dict__)
 
-        self.node = self.rank // self.ranks_per_node
-        self.num_nodes = self.num_ranks // self.ranks_per_node
-        self.is_multinode = 1 < self.num_nodes
+    @classmethod
+    def _get_worker_info(cls) -> Tuple[int, int]:
+        """Get worker info, or default to 0 of 1.
 
+        Returns:
+            Tuple[int, int]: Worker ID out of how many workers.
+        """
         info = get_worker_info()
         if info:
-            self.worker_of_rank = info.id
-            self.workers_per_rank = info.num_workers
+            ret = info.id, info.num_workers
         else:
-            self.worker_of_rank = 0
-            self.workers_per_rank = 1
-
-        self.worker = self.rank * self.workers_per_rank + self.worker_of_rank
-        self.num_workers = self.num_ranks * self.workers_per_rank
-        self.worker_of_node = self.rank_of_node * self.workers_per_rank + self.worker_of_rank
-        self.workers_per_node = self.ranks_per_node * self.workers_per_rank
-
-        self.is_leader = not self.worker
-        self.is_local_leader = not self.worker_of_node
+            ret = 0, 1
+        return ret
+
+    @classmethod
+    def detect(cls) -> Self:
+        """Detect the world state.
+
+        Returns:
+            Self: A new World state object according to dist and get_worker_info().
+        """
+        rank = dist.get_rank()
+        ranks_per_node = dist.get_local_world_size()
+        num_nodes = dist.get_world_size() // ranks_per_node
+        worker_of_rank, workers_per_rank = cls._get_worker_info()
+        worker = rank * workers_per_rank + worker_of_rank
+        return cls(num_nodes, ranks_per_node, workers_per_rank, worker)
+
+    def tensor_parallel(self, ratio: int) -> Self:
+        """Get a copy of this world state with the given tensor paralellism.
+
+        Args:
+            ratio (int): Ratio of tensor parallelism.
+
+        Returns:
+            Self: A new tensor-parallel version of this World state object.
+        """
+        if 0 <= ratio:
+            raise ValueError(f'Tensor parallelism ratio must be postiive.')
+
+        if self.num_ranks % ratio:
+            raise ValueError(f'World size must be divisible by your tensor parallelism ratio.')
+
+        rank = self.rank // ratio
+        num_ranks = self.num_ranks // ratio
+        worker = rank * self.workers_per_rank + self.worker_of_rank
+        if self.ranks_per_node <= num_ranks:
```

Comment:
@knighton mind adding some comments to the code here? I'm trying to understand what modifications are happening, and I _think_ there may be a bug or a better way of doing this, but it's pretty hard for me to reason about without completely understanding what you're trying to accomplish here.

iiuc `ratio` is the number of consecutive GPUs that should be sharing the same samples -- and should have the same `World` information. `num_ranks` here is the # of TP blocks, and `rank` is the TP block index. I'm not seeing why we need to check if `self.ranks_per_node <= num_ranks`, and it seems if we set `num_nodes` to 1, then we'll have download duplication across all the nodes which will be pretty bad. The TP ratio should be the degree of duplication of the `World` objects and I'm not sure that this does that entirely correctly...

Reply:
> iiuc ratio is the number of consecutive GPUs that should be sharing the same samples -- and should have the same World information. num_ranks here is the # of TP blocks, and rank is the TP block index.

Exactly (for iterating purposes, modulo that bug for coordinating purposes which am about to fix)

> I'm not seeing why we need to check if `self.ranks_per_node <= num_ranks`, and it seems if we set num_nodes to 1, then we'll have download duplication across all the nodes which will be pretty bad.

PR originally applied tensor parallelism intra-node only, but then iiuc VC noted TP blocks may be inter-node here: https://github.com/mosaicml/streaming/pull/597#discussion_r1484550537

Reply:
Spelling out the logic for my own benefit:

```
If num TP blocks exceeds true local world size:
    Scale down perceived num nodes to TP blocks / true local world size
    Don't need to scale down its perceived local world size because we scaled nodes
else:
    Just one perceived node, and what happens in the node, stays in the node
    Map ranks to potentially fewer TP blocks by potentially scaling down its peceived local world size
```

Reply:
Could rewrite it to:

```
num_nodes = (num_ranks + self.ranks_per_node - 1) // self.ranks_per_node
ranks_per_node = num_ranks // num_nodes
```

Reply:
> modulo that bug for coordinating purposes which am about to fix

https://github.com/mosaicml/streaming/pull/597/commits/a17419eafbac1669e02779511c5e2470c4da1758